src/main/java/com/example/medicalclinic/service/DoctorService.java

```diff
@@ -0,0 +1,116 @@
+                .map(doctorMapper::toDTO)
+                .toList();
+
+        return PageableContentDTO.from(doctorPage, doctorDTOS);
+    }
+
+    public DoctorDTO getDoctorByEmail(String email) {
+        return doctorMapper.toDTO(findDoctorByEmail(email));
+    }
+
+    @Transactional
+    public Doctor addDoctor(Doctor doctor) {
+        if (doctorRepository.findByEmail(doctor.getEmail()).isPresent()) {
+            throw new DoctorException("Doctor with email: " + doctor.getEmail() + " already exists");
+        }
+        return doctorRepository.save(doctor);
+    }
+
+    @Transactional
+    public void removeDoctorByEmail(String email) {
+        Doctor doctor = findDoctorByEmail(email);
+        doctorRepository.delete(doctor);
+    }
+
+    public DoctorDTO editDoctorByEmail(String email, Doctor updatedDoctor) {
+        Doctor existingDoctor = findDoctorByEmail(email);
+        updateEmailIfChanged(existingDoctor, updatedDoctor);
+        existingDoctor.updateFrom(updatedDoctor);
+        return doctorMapper.toDTO(doctorRepository.save(existingDoctor));
+    }
+
+    public Doctor changePassword(String email, String password) {
+        Doctor existingDoctor = findDoctorByEmail(email);
+        existingDoctor.setPassword(password);
+        return doctorRepository.save(existingDoctor);
+    }
+
+    @Transactional
+    public DoctorDTO assignDoctorToFacility(Long doctorId, Long facilityId) {
+        Doctor doctor = findDoctorById(doctorId);
+        Facility facility = findFacilityById(facilityId);
+
+        if (!doctor.getFacilities().contains(facility)) {
+            doctor.getFacilities().add(facility);
+        }
+
+        return doctorMapper.toDTO(doctorRepository.save(doctor));
+    }
+
+    @Transactional
+    public void removeFacilityFromDoctor(Long doctorId, Long facilityId) {
+        Doctor doctor = findDoctorById(doctorId);
+        Facility facility = findFacilityById(facilityId);
+
+        if (doctor.getFacilities().contains(facility)) {
+            doctor.getFacilities().remove(facility);
+            doctorRepository.save(doctor);
+        }
+    }
+
+    private Doctor findDoctorByEmail(String email) {
+        return doctorRepository.findByEmail(email)
+                .orElseThrow(() -> new DoctorException("Doctor with email: " + email + " not found"));
+    }
+
+    private Doctor findDoctorById(Long id) {
+        return doctorRepository.findById(id)
+                .orElseThrow(() -> new DoctorException("Doctor with ID: " + id + " not found"));
+    }
+
+    private Facility findFacilityById(Long facilityId) {
+        return facilityRepository.findById(facilityId)
+                .orElseThrow(() -> new FacilityException("Facility with ID: " + facilityId + " not found"));
+    }
+
+    private void updateEmailIfChanged(Doctor existingDoctor, Doctor updatedDoctor) {
+        String newEmail = updatedDoctor.getEmail();
+        if (newEmail != null && !newEmail.equals(existingDoctor.getEmail())) {
```

Comment:
wyniosl ten warunek do oddzielnego booleana albo utworzyl predicate w tej metodzie ktory by opisywał co to robi